pkg/querier/remote_read_test.go

```diff
@@ -1244,3 +1244,135 @@ func TestQueryFromRemoteReadQuery(t *testing.T) {
+			return &mockQuerier{
+				selectFn: func(ctx context.Context, sorted bool, hints *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet {
+					concurrentQueries.Inc()
+					defer concurrentQueries.Dec()
+					<-controlChan
+
+					// Return a simple series set
+					return series.NewConcreteSeriesSetFromUnsortedSeries(
+						[]storage.Series{
+							series.NewConcreteSeries(labels.FromStrings("foo", "bar"), []model.SamplePair{{Timestamp: 1, Value: 1.0}}, nil),
+						},
+					)
+				},
+			}, nil
+		},
+	}
+
+	tests := []struct {
+		name                  string
+		queries               int
+		maxConcurrency        int
+		expectedMaxConcurrent int32
+	}{
+		{
+			name:                  "unlimited concurrency (0) allows all queries to run concurrently",
+			queries:               5,
+			maxConcurrency:        0,
+			expectedMaxConcurrent: 5,
+		},
+		{
+			name:                  "concurrency limit of 2 restricts to 2 concurrent queries",
+			queries:               5,
+			maxConcurrency:        2,
+			expectedMaxConcurrent: 2,
+		},
+		{
+			name:                  "concurrency limit of 1 serializes all queries",
+			queries:               3,
+			maxConcurrency:        1,
+			expectedMaxConcurrent: 1,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Reset counters and ensure control channel is empty
+			concurrentQueries.Store(0)
+			for len(controlChan) > 0 {
+				<-controlChan
+			}
+
+			// Create handler with configurable concurrency
+			handler := RemoteReadHandler(q, log.NewNopLogger(), Config{MaxConcurrentRemoteReadQueries: tt.maxConcurrency})
+
+			// Create multiple queries
+			queries := make([]*prompb.Query, tt.queries)
+			for i := 0; i < tt.queries; i++ {
+				queries[i] = &prompb.Query{
+					StartTimestampMs: 1,
+					EndTimestampMs:   10,
+					Matchers: []*prompb.LabelMatcher{
+						{Type: prompb.LabelMatcher_EQ, Name: labels.MetricName, Value: "test_metric"},
+					},
+				}
+			}
+
+			requestBody, err := proto.Marshal(&prompb.ReadRequest{Queries: queries})
+			require.NoError(t, err)
+			requestBody = snappy.Encode(nil, requestBody)
+
+			request, err := http.NewRequest(http.MethodPost, "/api/v1/read", bytes.NewReader(requestBody))
+			require.NoError(t, err)
+			request.Header.Add("Content-Encoding", "snappy")
+			request.Header.Set("Content-Type", "application/x-protobuf")
+
+			response := httptest.NewRecorder()
+
+			// Start request in goroutine
+			done := make(chan struct{})
+			go func() {
+				defer close(done)
+				handler.ServeHTTP(response, request)
+			}()
+
+			// Wait for expected concurrency to be reached and verify limits
+			maxObserved := int32(0)
+			require.Eventually(t, func() bool {
+				current := concurrentQueries.Load()
+				if current > maxObserved {
+					maxObserved = current
+				}
+
+				// Check that we never exceed the expected limit
+				require.LessOrEqual(t, current, tt.expectedMaxConcurrent,
```

Comment:
Does this need to be this?

```suggestion
				require.LessOrEqualf(t, current, tt.expectedMaxConcurrent,
```